### PR TITLE
Replace std::sort with gfx::timsort in Spiller for better performance

### PR DESCRIPTION
**Pull request by yma11**
Based on profile of workloads that trigger spills, `std::sort` is the key
hotspot in spilling of HashAggregation operator.

Replace std::sort with gfx::timsort from https://github.com/timsort/cpp-TimSort. 

This change provides >2X performance win.

### CODE REVIEW

**Comment by yma11** [comment]
> Can you add a benchmark comparing `timsort` with `std::sort` on different distributions?

Do you mean add a test based on Velox benchmark framework? For the perf data on different distributions, you can take a reference first from [Here](https://github.com/scandum/quadsort#benchmark-quadsort-vs-stdstable_sort-vs-timsort).  According to the benchmarks, gfx::timsort is slower than [std::sort()](https://en.cppreference.com/w/cpp/algorithm/sort) on randomized sequences, but faster on partially-sorted ones, which is more complied with real-world data. Timsort is widely used in frameworks like Spark, Pandas, Python as well as Java Array.sort(), etc.

**Review by mbasmanova — COMMENTED** [review]
@yma11 Thank you for the optimization. I'm still reading the PR, but have a quick request.

Please, move TimSort implementation into velox/external.

File: velox/exec/TimSort.hpp [hunk 1]
```diff
@@ -0,0 +1,814 @@
+/*
```
Comment (xiaoxmeng) [inline velox/exec/TimSort.hpp:1]: Can we have any basic testing on this utility? Also a benchmark to show the performance difference? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@yma11 thanks for the optimization. LGTM. Can we have a folly benchmark for this?

**Comment by laithsakka** [comment]
"according to the benchmarks, gfx::timsort is slower than [std::sort()](https://en.cppreference.com/w/cpp/algorithm/sort) on randomized sequences, but faster on partially-sorted ones. It can be used as a drop-in replacement for [std::stable_sort](https://en.cppreference.com/w/cpp/algorithm/stable_sort), with the difference that it can't fallback to a O(n log² n) algorithm when there isn't enough extra heap memory available."

is that a static property of spilled workload?

**Comment by Yuhta** [comment]
For spilling, in memory sort probably is not going to be the bottleneck and the main optimization should be reduce IO

**Comment by Yuhta** [comment]
I would suggest we wait until seeing a real world use case to use this new sort instead of applying it blindly everywhere.

**Comment by yma11** [comment]
> For spilling, in memory sort probably is not going to be the bottleneck and the main optimization should be reduce IO

We profiled a case like query `select key_1, key_2, sum(key_3), sum(key_4), ...(more aggs) from table grou by  key_1, key_2` using customer data , in which partial aggregation doesn't reduce rows and heavy spills happen in final phase. The `Spiller::writeSpill()` takes 65.1% of total computation in final agg and in which `ensureSorted()` takes 63.7%, means sort occupies 97.8% when do spill. Recently spill in hashAgg is optimized but sort vs spill total time should not change.

**Comment by yma11** [comment]
> @yma11 The change looks good. Would it be possible to include a benchmark that shows 2x performance win?

Yes. Working on that.

**Comment by Yuhta** [comment]
Let's add it to Spiller first then.  We will also need a query & data setup (can be described in the PR summary) to reproduce that performance gain so we can reuse the case when doing future optimization

**Comment by yma11** [comment]
@mbasmanova @Yuhta @xiaoxmeng This PR is update 1) limit use of timsort only in `Spiller`, 2) add benchmarks on different data patterns generated leveraging velox test utility. Please help review again. The benchmark result is shown as following:
```
============================================================================
[...]chmarks/SimpleVectorSortBenchmark.cpp     relative  time/iter   iters/s
============================================================================
BM_Int64_stdSort(100k_rows_100k_uni_noseq)                  7.16ms    139.68
BM_Int64_timSort(100k_rows_100k_uni_noseq)      80.026%     8.95ms    111.78
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_10k_uni_noseq)                   6.87ms    145.64
BM_Int64_timSort(100k_rows_10k_uni_noseq)       83.902%     8.18ms    122.19
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_1k_uni_noseq)                    5.57ms    179.62
BM_Int64_timSort(100k_rows_1k_uni_noseq)        79.074%     7.04ms    142.03
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_10k_uni_seq)                     2.20ms    454.79
BM_Int64_timSort(100k_rows_10k_uni_seq)         856.37%   256.76us     3.89K
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_1k_uni_seq)                      2.14ms    466.52
BM_Int64_timSort(100k_rows_1k_uni_seq)          433.64%   494.30us     2.02K
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_10k_uni_noseq)                    72.02ms     13.89
BM_Int64_timSort(1M_rows_10k_uni_noseq)          77.63%    92.77ms     10.78
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_1k_uni_noseq)                     60.31ms     16.58
BM_Int64_timSort(1M_rows_1k_uni_noseq)          78.493%    76.84ms     13.01
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_10k_uni_seq)                      27.73ms     36.07
BM_Int64_timSort(1M_rows_10k_uni_seq)           309.72%     8.95ms    111.70
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_1k_uni_seq)                       27.36ms     36.55
BM_Int64_timSort(1M_rows_1k_uni_seq)            222.91%    12.27ms     81.47
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_100_uni_seq)                      26.99ms     37.06
BM_Int64_timSort(1M_rows_100_uni_seq)           283.97%     9.50ms    105.23
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_10k_uni_noseq)                   146.71ms      6.82
BM_Int64_timSort(2M_rows_10k_uni_noseq)          71.44%   205.36ms      4.87
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_1k_uni_noseq)                    121.21ms      8.25
BM_Int64_timSort(2M_rows_1k_uni_noseq)          69.568%   174.23ms      5.74
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_10k_uni_seq)                      59.68ms     16.76
BM_Int64_timSort(2M_rows_10k_uni_seq)           184.55%    32.34ms     30.92
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_1k_uni_seq)                       56.84ms     17.59
BM_Int64_timSort(2M_rows_1k_uni_seq)            152.11%    37.37ms     26.76
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_100_uni_seq)                      57.15ms     17.50
BM_Int64_timSort(2M_rows_100_uni_seq)           161.73%    35.34ms     28.30
----------------------------------------------------------------------------

```

**Review by mbasmanova — COMMENTED** [review]
@yma11 Thank you for iterating on this PR. The code changes look good, but I'm not sure I understand the benchmark. It seems to be comparing timsort vs. std::sort and that comparison is available in timsort's repo already, no? I was expecting a benchmark that would sort Velox vectors of different types to match spilling use case, not std::vector. Am I missing something?

**Comment by mbasmanova** [comment]
I'm also seeing timsort used in Cosco shuffle.

I suggest to remove the benchmark from this PR and merge it.

**Comment by mbasmanova** [comment]
@yma11 Curious if you could maybe share some of the profiles of e2e queries you collected before and after this change.

**Comment by Yuhta** [comment]
@yma11 So the result is very clear that timsort sort is good at sorting presorted sequences.  I am curious how sorted the data is required to be for timsort to beat std::sort (in terms of average sequence length / total length), do you have this data?

@xiaoxmeng Is spilling data mainly consists of sorted sequences?

**Comment by xiaoxmeng** [comment]
> @yma11 So the result is very clear that timsort sort is good at sorting presorted sequences. I am curious how sorted the data is required to be for timsort to beat std::sort (in terms of average sequence length / total length), do you have this data?
> 
> @xiaoxmeng Is spilling data mainly consist of sorted sequences?

@yma11 @Yuhta spilling data is not pre-sorted.

**Comment by yma11** [comment]
> @yma11 Thank you for iterating on this PR. The code changes look good, but I'm not sure I understand the benchmark. It seems to be comparing timsort vs. std::sort and that comparison is available in timsort's repo already, no? I was expecting a benchmark that would sort Velox vectors of different types to match spilling use case, not std::vector. Am I missing something?

`std::sort` and `gfx::timsort` need take an `IteratorableRow` as input for sort but Velox vector doesn't have such interface. I updated it to benchmark in the `RowContainer` instead, in which 1) generate the data 2) store the data in `RowContainer`  3) sort on rows in container. This simulates the same code as `ensureSorted()` during spill. The new data shows that `timsort` has perf gain even in random sequence when the cardinality is small. I will SOON update this PR with new data. Stay tuned.

**Comment by yma11** [comment]
> @yma11 Curious if you could maybe share some of the profiles of e2e queries you collected before and after this change.

As mentioned before, the query we profiled is like `select key_1, key_2, sum(key_3), sum(key_4), ...(more aggs) from table grou by  key_1, key_2`. In this query, `key_1`, `key_2` are the primary keys so in this query, partial agg doesn't reduce input rows and then cause heavy spill in final hashAgg, which happens in both vanilla spark and velox. The e2e time shows that velox takes 2X longer even the spilled data size is much smaller which indicates spill's low efficiency. After switching to timesort, velox got a comparable perf vs spark.
```
                                                 vanilla spark                                 velox-std::sort                            velox-timsort
e2e time                                         57s                                                137s                                           60s
spilled data size                            6.1 GiB                                          342.4 MiB
```
Here is the vtune data which shows the hotspots.
`before`
image

`after `
image

So in summary, you can see the sort time drops from 137s*0.64=87.7s to 60s*0.46=27.6s after switching to timsort.

**Comment by yma11** [comment]
> @yma11 So the result is very clear that timsort sort is good at sorting presorted sequences. I am curious how sorted the data is required to be for timsort to beat std::sort (in terms of average sequence length / total length), do you have this data?
> 
> @xiaoxmeng Is spilling data mainly consist of sorted sequences?

@Yuhta [here](https://www.baeldung.com/cs/timsort) is an article about how timsort works. The sorted subsequence in the data will help a lot during insert sort phase in timsort. 
image

**Comment by yma11** [comment]
> > @yma11 Thank you for iterating on this PR. The code changes look good, but I'm not sure I understand the benchmark. It seems to be comparing timsort vs. std::sort and that comparison is available in timsort's repo already, no? I was expecting a benchmark that would sort Velox vectors of different types to match spilling use case, not std::vector. Am I missing something?
> 
> `std::sort` and `gfx::timsort` need take an `IteratorableRow` as input for sort but Velox vector doesn't have such interface. I updated it to benchmark in the `RowContainer` instead, in which 1) generate the data 2) store the data in `RowContainer` 3) sort on rows in container. This simulates the same code as `ensureSorted()` during spill. The new data shows that `timsort` has perf gain even in random sequence when the cardinality is small. I will SOON update this PR with new data. Stay tuned.

@mbasmanova benchmark based on `RowContainer` updated and here is the new perf data. See the ones highlighted with `>>>>>>>>`.

```
============================================================================
[...]chmarks/RowContainerSortBenchmark.cpp     relative  time/iter   iters/s
============================================================================
BM_Int64_stdSort(100k_rows_100k_uni_noseq)                 27.74ms     36.05
BM_Int64_timSort(100k_rows_100k_uni_noseq)      95.767%    28.96ms     34.53
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_10k_uni_noseq)                  26.55ms     37.66
BM_Int64_timSort(100k_rows_10k_uni_noseq)       97.243%    27.30ms     36.63
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_1k_uni_noseq)                   23.33ms     42.86
>>>>>>>> BM_Int64_timSort(100k_rows_1k_uni_noseq)        102.19%    22.83ms     43.80
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_10k_uni_seq)                    14.63ms     68.37
BM_Int64_timSort(100k_rows_10k_uni_seq)         1504.1%   972.36us     1.03K
----------------------------------------------------------------------------
BM_Int64_stdSort(100k_rows_1k_uni_seq)                     13.48ms     74.20
BM_Int64_timSort(100k_rows_1k_uni_seq)            1156%     1.17ms    857.70
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_10k_uni_noseq)                   315.20ms      3.17
>>>>>>>> BM_Int64_timSort(1M_rows_10k_uni_noseq)         102.99%   306.03ms      3.27
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_1k_uni_noseq)                    289.12ms      3.46
>>>>>>>> BM_Int64_timSort(1M_rows_1k_uni_noseq)          126.77%   228.07ms      4.38
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_10k_uni_seq)                     187.58ms      5.33
BM_Int64_timSort(1M_rows_10k_uni_seq)           1436.1%    13.06ms     76.56
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_1k_uni_seq)                      185.05ms      5.40
BM_Int64_timSort(1M_rows_1k_uni_seq)            1075.7%    17.20ms     58.13
----------------------------------------------------------------------------
BM_Int64_stdSort(1M_rows_100_uni_seq)                     180.56ms      5.54
BM_Int64_timSort(1M_rows_100_uni_seq)           1365.1%    13.23ms     75.60
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_10k_uni_noseq)                   685.67ms      1.46
>>>>>>>> BM_Int64_timSort(2M_rows_10k_uni_noseq)         108.67%   630.95ms      1.58
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_1k_uni_noseq)                    612.26ms      1.63
>>>>>>>> BM_Int64_timSort(2M_rows_1k_uni_noseq)          131.28%   466.39ms      2.14
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_10k_uni_seq)                     394.87ms      2.53
BM_Int64_timSort(2M_rows_10k_uni_seq)           1200.1%    32.90ms     30.39
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_1k_uni_seq)                      373.38ms      2.68
BM_Int64_timSort(2M_rows_1k_uni_seq)            1024.5%    36.44ms     27.44
----------------------------------------------------------------------------
BM_Int64_stdSort(2M_rows_100_uni_seq)                     382.30ms      2.62
BM_Int64_timSort(2M_rows_100_uni_seq)           1159.1%    32.98ms     30.32
----------------------------------------------------------------------------

```

**Comment by mbasmanova** [comment]
@yma11 

> The new data shows that timsort has perf gain even in random sequence when the cardinality is small.

What do you mean by "cardinality is small"? If RowContainer has only few entries then we won't need spilling at all. Am I missing something?

**Comment by Yuhta** [comment]
> @yma11
> 
> > The new data shows that timsort has perf gain even in random sequence when the cardinality is small.
> 
> What do you mean by "cardinality is small"? If RowContainer has only few entries then we won't need spilling at all. Am I missing something?

I guess he means there are only a few unique keys (thus same contiguous keys form longer sequences as a result)

**Comment by Yuhta** [comment]
> > @yma11 Curious if you could maybe share some of the profiles of e2e queries you collected before and after this change.
> 
> As mentioned before, the query we profiled is like `select key_1, key_2, sum(key_3), sum(key_4), ...(more aggs) from table grou by key_1, key_2` where `key_1`, `key_2` are the primary keys. So in this query, partial agg doesn't reduce input rows and then cause heavy spill in final hashAgg, which happens in both vanilla spark and velox. The e2e time shows that velox takes 2X longer even the spilled data size is much smaller which indicates spill's low efficiency. After switching to timesort, velox got a comparable perf vs spark.
> 
> ```
>                                                  vanilla spark                                 velox-std::sort                            velox-timsort
> e2e time                                         57s                                                137s                                           60s
> spilled data size                            6.1 GiB                                          342.4 MiB
> ```
> 
> Here is the vtune data which shows the hotspots. `before` image
> 
> `after ` image
> 
> So in summary, you can see the sort time drops from 137sx0.64=87.7s to 60sx0.46=27.6s after switching to timsort.

So in this experiment all input keys are unique?

**Comment by mbasmanova** [comment]
@yma11 Would you confirm whether the input data is pre-sorted or whether you use sorted shuffle?

**Comment by yma11** [comment]
> ```
>                                                  vanilla spark                                 velox-std::sort                            velox-timsort
> e2e time                                         57s                                                137s                                           60s
> spilled data size                            6.1 GiB                                          342.4 MiB
> ```
> 
> @yma11 I wonder why Spark is spilling so much more data (6GB vs. 342 MB). Do you think spilled data size reporting in Velox is wrong? Or is there something else going on? With Velox spilling 17x less data and spilling being the perf bottleneck, I'd expect velox performance to be a lot better.
In our practice, the spill size difference is reasonable as Spark has much higher memory consumption than Velox and will trigger more data spill. But Spark has better sort strategy, probably used just a Radix sort against Long Array. 
> 
> I see HashTable::erase being very high on the profile. I discussed this with @xiaoxmeng yesterday and we think we should remove this code altogether from spilling.

Yes. I shared the `eraseRow()` with him that should be optimized as next step. In the hashAgg spilling called by reclaimer, the whole hashtable should can be cleared at one time.

**Comment by yma11** [comment]
> > @yma11
> > > The new data shows that timsort has perf gain even in random sequence when the cardinality is small.
> > 
> > 
> > What do you mean by "cardinality is small"? If RowContainer has only few entries then we won't need spilling at all. Am I missing something?
> 
> I guess he means there are only a few unique keys (thus same contiguous keys form longer sequences as a result)

Yes. `BM_Int64_timSort(2M_rows_1k_uni_noseq` means we have 2M rows total with 1k unique values.

**Comment by yma11** [comment]
> So in this experiment all input keys are unique? How do you decide the order of the keys in input data?

Yes. The input keys are all unique in this customer workload thus the partial Agg doesn't reduce rows at all. These keys are strings auto generated by other workloads. I can not tell whether they are strictly ordered as selected rows from the table. But based on the perf data, it should only have some ordered subsequence or perf improvement should be higher.

**Comment by Yuhta** [comment]
> > So in this experiment all input keys are unique? How do you decide the order of the keys in input data?
> 
> Yes. The input keys are all unique in this customer workload thus the partial Agg doesn't reduce rows at all. These keys are strings auto generated by other workloads. I can not tell whether they are strictly ordered as selected rows from the table. But based on the perf data, it should only have some ordered subsequence or perf improvement should be higher.

Can you calculate the average sorted sequence length and total length for the raw input, if it does not take too much work on your side?  The performance of timsort heavily depends on this parameter

**Comment by yma11** [comment]
> > > So in this experiment all input keys are unique? How do you decide the order of the keys in input data?
> > 
> > 
> > Yes. The input keys are all unique in this customer workload thus the partial Agg doesn't reduce rows at all. These keys are strings auto generated by other workloads. I can not tell whether they are strictly ordered as selected rows from the table. But based on the perf data, it should only have some ordered subsequence or perf improvement should be higher.
> 
> Can you calculate the average sorted sequence length and total row count for the raw input, if it does not take too much work on your side? The performance of timsort heavily depends on this parameter

Do you mean `runLength` that doing merge sort against? We don't tune that parameter but use the one as the algorithm calculated.

**Comment by Yuhta** [comment]
> Do you mean `runLength` that doing merge sort against? We don't tune that parameter but use the one as the algorithm calculated.

Yes this is a characteristic of data and cannot be controlled in real world.  Would like to have it known for test data so that future benchmarks can reproduce this result.

**Comment by yma11** [comment]
> > Do you mean `runLength` that doing merge sort against? We don't tune that parameter but use the one as the algorithm calculated.
> 
> Yes this is a characteristic of data and cannot be controlled in real world. Would like to have it known for test data so that future benchmarks can reproduce this result.

@Yuhta  I printed the `runLength` in each iteration and found the values are quite small like `3, 2, 2, 3, 4, 2, 5 ..., 24, 4, 2, 3 .....` with total `336321` iterations. At the last iteration, the value becomes `1717022` which is quite large. The total sorted row number is `28,319,541`. Seems this is a quite common data pattern.

**Comment by Yuhta** [comment]
> > > Do you mean `runLength` that doing merge sort against? We don't tune that parameter but use the one as the algorithm calculated.
> > 
> > 
> > Yes this is a characteristic of data and cannot be controlled in real world. Would like to have it known for test data so that future benchmarks can reproduce this result.
> 
> @Yuhta I printed the `runLength` in each iteration and found the values are quite small like `3, 2, 2, 3, 4, 2, 5 ..., 24, 4, 2, 3 .....` with total `336321` iterations. At the last iteration, the value becomes `1717022` which is quite large. The total sorted row number is `28,319,541`. Seems this is a quite common data pattern.

Are you print the runLength during the divide & conquer process of the sorting?  Then it is expected that it becomes progressively larger.  What I am interested in is the average of the run lengths in the raw input data.  This should be very easy to calculate in a single pass of the data using about 5 lines of code.

IID sequence from uniform distribution has this average run length at (e - 1): https://stats.stackexchange.com/questions/350923/brain-teaser-what-is-the-expected-length-of-an-iid-sequence-that-is-monotonical, apparently this is where `std::sort` wins.  Curious what is the minimum value makes `timsort` wins.

**Comment by yma11** [comment]
> Are you print the runLength during the divide & conquer process of the sorting? Then it is expected that it becomes progressively larger. What I am interested in is the average of the run lengths in the raw input data. This should be very easy to calculate in a single pass of the data using about 5 lines of code.
> 
> IID sequence from uniform distribution has this average run length at (e - 1): https://stats.stackexchange.com/questions/350923/brain-teaser-what-is-the-expected-length-of-an-iid-sequence-that-is-monotonical, apparently this is where `std::sort` wins. Curious what is the minimum value makes `timsort` wins.

@Yuhta In this workload, there are 2 group keys which means the layout of row in `RowContainer` is `key_1, key_2, sum_1, sum_2......` where `key_1` and `key_2` are string values. So I did an experiment to calculate the avg sorted length against a merge of columns `key_1` and `key_2`, by `total sorted sequence length`/`total sorted sequence number`, considering both `asc` and `dsc` orders.
```
  auto query_sig_col =
      data->as<RowVector>()->childAt(0)->asFlatVector<StringView>();
  auto result_sig_col =
      data->as<RowVector>()->childAt(1)->asFlatVector<StringView>();
  // Get a vector of order, true means current value bigger than previous.
  // This vectors turns out to be like "10101001010...."
  std::vector<bool> sorted(28319540, false);
  auto merge_col_pre = query_sig_col->valueAt(0).getString() +
      result_sig_col->valueAt(0).getString();
  for (int i = 1; i < query_sig_col->size(); i++) {
    auto merge_col = query_sig_col->valueAt(i).getString() +
        result_sig_col->valueAt(i).getString();
    if (merge_col < merge_col_pre) {
      sorted[i - 1] = false;
    } else if (merge_col == merge_col_pre && i > 1) {
      sorted[i - 1] = sorted[i - 2];
    } else if (merge_col == merge_col_pre && i == 1) {
      // Unnecessary handle for test data as the first 2 rows are different
      sorted[i - 1] = true;
    } else if (merge_col > merge_col_pre) {
      sorted[i - 1] = true;
    }
    merge_col_pre = merge_col;
  }
  // Calculate the avg sorted length 
  int asc_seq_num = 0, asc_count = 0, dsc_seq_num = 0, dsc_count = 0;
  for (int i = 1; i < query_sig_col->size(); i++) {
    if (sorted[i]) {
      asc_count++;
      if (!sorted[i - 1]) {
        asc_seq_num++;
      }
    }
    if (!sorted[i]) {
      dsc_count++;
      if(sorted[i-1]) {
        dsc_seq_num++;
      }
    }
  }
  std::cout << (float)(asc_count+dsc_count)/(asc_seq_num+dsc_seq_num) << std::endl;
```

The value turns out to be  `1.00947` which seems comply with the `runLength` print which shows no large preordered sequence in this data. 
I think perf improvement of `timsort` is obvious based on benchmark on `RowContainer`.  It's already comparable (95%, 97%) to `std::sort` in the worst case `100k_rows_100k_uni_noseq` and have gains in all other cases.

**Comment by Yuhta** [comment]
@yma11 1.00947 is very surprising as it means you have data that is almost zigzag on each row.  I think even uniform distribution should perform better than this for `timsort` (I don't see this effect in RowContainer benchmark though, which has 95% performance of std::sort).  Either way it should be good for Spiller.

**Comment by yma11** [comment]
@mbasmanova @xiaoxmeng @Yuhta Any more comments for this change? Thanks.

File: velox/exec/benchmarks/RowContainerSortBenchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+namespace facebook::velox::test {
+namespace {
+
+template <typename T>
+VectorGeneratedData<T>
+testData(size_t length, size_t cardinality, bool sequences, size_t iteration) {
+  int32_t seqCount =
+      sequences ? std::max(2, (int32_t)(length / cardinality)) : 0;
+  int32_t seqLength = sequences ? std::max(2, (int32_t)(length / seqCount)) : 0;
+  return genTestDataWithSequences<T>(
```
Comment (Yuhta) [inline velox/exec/benchmarks/RowContainerSortBenchmark.cpp:36]: One more thing here, `genTestDataWithSequences` is generating same value sequences, so even when you disable the sequence, it can still form a monotonic increasing/decreasing sequence.  Can you replace this with something that we can have more control over the monotonic increasing/decreasing sequence length?

Comment (yma11) [reply velox/exec/benchmarks/RowContainerSortBenchmark.cpp:36]: @Yuhta I subsampled 500 rows data from the set that I get `runLength = 1.00947`. The perf is:
image

One open is that I should remove the previous benchmark leveraging `genTestDataWithSequences `, how do you think?

Comment (Yuhta) [reply velox/exec/benchmarks/RowContainerSortBenchmark.cpp:36]: Can you sample 50k rows?
For the sequence I think you can just use `genTestData` and just benchmark the uniform distribution case.  So we will have the parquet file to represent real world data and `genTestData` for worst case random data and I think that's enough for now.

Comment (yma11) [reply velox/exec/benchmarks/RowContainerSortBenchmark.cpp:39]: @Yuhta PR updated as you suggest. Please help take a review again and approve it if there is no further problem. Thanks.

Comment (yma11) [reply velox/exec/benchmarks/RowContainerSortBenchmark.cpp]: @Yuhta Switched to `genTestData` and remove redundant benchmark since now only `cardinality` matters. The latest data is:
image
FYI.

**Comment by mbasmanova** [comment]
FYI, https://danlark.org/2022/04/20/changing-stdsort-at-googles-scale-and-beyond/
